packages/react-query/src/shared/hooks/createHooksInternal.tsx

```diff
@@ -181,7 +182,36 @@ export function createRootHooks<
                 },
               };
 
-              return client.query(...getClientArgs(queryKey, actualOpts));
+              const result = await client.query(
+                ...getClientArgs(queryKey, actualOpts),
+              );
+
+              if (isAsyncIterable(result)) {
+                const queryHash = hashKey(queryKey);
+
+                const queryCache = queryClient.getQueryCache();
+
+                const query = queryCache.build(queryClient, {
+                  queryKey,
+                  queryHash,
+                });
+
+                query.setState({
+                  data: [],
+                  status: 'pending',
+                });
+
+                const aggregate: unknown[] = [];
+                for await (const value of result) {
+                  aggregate.push(value);
+
+                  query.setState({
+                    data: [...aggregate],
+                  });
+                }
+                return aggregate;
```

Comment:
Maybe we shouldn't mutate the state once it's be handed over to tanstack/query
```suggestion
                let aggregate: unknown[] = [];
                for await (const value of result) {
                  aggregate = [...aggregate, value];

                  query.setState({
                    data: aggregate,
                  });
                }
                return aggregate;
```

Reply:
That's why I already make a copy when setting the state like

```ts
query.setState({
  data: [...aggregate],
});
```

Reply:
fair enough, i must have misread